File: 261. Graph Valid Tree.cpp
//Solution1 (DFS)
class Solution {
public:
    bool validTree(int n, vector<pair<int, int>>& edges) {
        vector<vector<int>> neighbors(n, vector<int>{});
        for (pair<int, int>& p : edges) {
            neighbors[p.first].push_back(p.second);
            neighbors[p.second].push_back(p.first);
        }
        vector<bool> visited(n, false);
        if (hasCycle(neighbors, 0, -1, visited))
            return false;
        for (bool v : visited) {
            if (!v) {
                return false;
            }
        }
        return true;
    }
    bool hasCycle(vector<vector<int>> &neighbors, int kid, int parent, vector<bool> &visited) {
        if (visited[kid]) return true;
        visited[kid] = true;
        for (auto &nbr : neighbors[kid]) {
            if (nbr != parent && hasCycle(neighbors, nbr, kid, visited))
                return true;
        }
        return false;
    }
};

//Solution2 (Union-Find)
class Solution {
private:
    int findParent(int cur, int parent[]) {
        while (cur != parent[cur]) {
            cur = parent[cur] = parent[parent[cur]];
        }
        return cur;
    }

public:
    bool validTree(int n, vector<pair<int, int>>& edges) {
        int parent[n], weight[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            weight[i] = 1;
        }
        for (auto e : edges) {
            int a = findParent(e.first, parent), b = findParent(e.second, parent);
            if (a == b) {
                return false;
            } else {
                if (weight[a] < weight[b]) { swap(a, b); }
                parent[b] = a, weight[a] += weight[b], n--;
            }
        }

        return n == 1;
    }
};

//Solution3: Another way to determine if cycle exists
class Solution {
public:
    bool validTree(int n, vector<pair<int, int> >& edges) {
        if (edges.size() != n - 1) return false;
        int component = 0;
        vector<int> v[n];
        vector<bool> visited(n, false);

        for (auto edge: edges) {
            v[edge.first].push_back(edge.second);
            v[edge.second].push_back(edge.first);
        }

        for (int i = 0; i < n ; i++) {
            if (!visited[i]) {
                ++component;
                DFS(v, visited, i);
            }
        }

        return component == 1;
    }

    void DFS(vector<int> v[], vector<bool>& visited, int index) {
        if (visited[index]) return;
        visited[index] = true;
        for (int j = 0; j < v[index].size(); ++j) {
            DFS(v, visited, v[index][j]);
        }
    }
};

File: java/261. Graph Valid Tree.java
class Solution {
    private class UnionFind {
        int[] parent;
        int[] weight;
        int count;
        UnionFind(int n) {
            parent = new int[n];
            weight = new int[n];
            count = n;
            for (int i = 0; i < n; ++i) {
                parent[i] = i;
            }
            Arrays.fill(weight, 1);
        }

        int find(int cur) {
            while (parent[cur] != cur) {
                parent[cur] = parent[parent[cur]];
                cur = parent[cur];
            }
            return cur;
        }

        boolean union(int nA, int nB) {
            int rootA = parent[nA];
            int rootB = parent[nB];
            if (rootA == rootB) {
                return false;
            }
            --count;
            if (weight[rootA] <= weight[rootB]) {
                parent[rootA] = rootB;
                weight[rootB] += weight[rootA];
            } else {
                parent[rootB] = rootA;
                weight[rootA] += weight[rootB];
            }
            return true;
        }
    }
    public boolean validTree(int n, int[][] edges) {
        UnionFind uf = new UnionFind(n);
        for (int i = 0; i < edges.length; ++i) {
            if (!uf.union(edges[i][0], edges[i][1])) {
                return false;
            }
        }
        return uf.count == 1;
    }
}
